File: SlopShopSparse/services/payments-csharp/Pagination.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SlopShop.Catalog
{
    public sealed class Page<T>
    {
        private const int DefaultSize = 20;
        private const int MaxSize = 100;

        public IReadOnlyList<T> Items { get; }
        public int Number { get; }
        public int Size { get; }
        public int Total { get; }

        private Page(IReadOnlyList<T> items, int number, int size, int total)
        {
            Items = items;
            Number = number;
            Size = size;
            Total = total;
        }

        public int Pages => Size == 0 ? 0 : (Total + Size - 1) / Size;

        public bool HasNext => Number < Pages;

        public bool HasPrevious => Number > 1;

        public static int NormalizeSize(int requested)
            => requested <= 0 ? DefaultSize : Math.Min(requested, MaxSize);

        public static Page<T> Of(IReadOnlyList<T> all, int number, int size)
        {
            number = Math.Max(1, number);
            size = NormalizeSize(size);
            var window = all.Skip((number - 1) * size).Take(size).ToList();
            return new Page<T>(window, number, size, all.Count);
        }


        /// <summary>1-based index of the first item on this page, or zero when empty.</summary>
        public int StartIndex => Items.Count == 0 ? 0 : (Number - 1) * Size + 1;

        /// <summary>1-based index of the last item on this page.</summary>
        public int EndIndex => (Number - 1) * Size + Items.Count;

        /// <summary>The "showing 21-40 of 137" line the listing views render.</summary>
        public string Describe()
            => Items.Count == 0 ? "no results" : $"showing {StartIndex}-{EndIndex} of {Total}";

        /// <summary>Row offset for a page, after both arguments are normalised.</summary>
        public static int OffsetFor(int number, int size)
            => (Math.Max(1, number) - 1) * NormalizeSize(size);

        /// <summary>Page numbers to render around the current one, clamped to range.</summary>
        public IEnumerable<int> Window(int span = 2)
        {
            if (Pages <= 0) yield break;
            var lo = Math.Max(1, Number - span);
            var hi = Math.Min(Pages, Number + span);
            for (var i = lo; i <= hi; i++) yield return i;
        }

        /// <summary>An empty but well-formed page, so callers never handle null.</summary>
        public static Page<T> Empty() => new Page<T>(Array.Empty<T>(), 1, DefaultSize, 0);
    }
}

File: SlopShopSparse/services/auth-python/pagination.py
"""Pagination utilities."""
from dataclasses import dataclass
from typing import List

DEFAULT_PAGE_SIZE = 20
MAX_PAGE_SIZE = 100


@dataclass
class Page:
    items: List
    page: int
    size: int
    total: int

    @property
    def pages(self) -> int:
        if self.size == 0:
            return 0
        return (self.total + self.size - 1) // self.size

    @property
    def has_next(self) -> bool:
        return self.page < self.pages

    @property
    def has_prev(self) -> bool:
        return self.page > 1

    def as_dict(self) -> dict:
        return {
            "page": self.page,
            "size": self.size,
            "total": self.total,
            "pages": self.pages,
            "has_next": self.has_next,
            "has_prev": self.has_prev,
        }

    @property
    def start_index(self) -> int:
        """1-based index of the first item on this page."""
        return 0 if not self.items else (self.page - 1) * self.size + 1

    @property
    def end_index(self) -> int:
        """1-based index of the last item on this page."""
        return (self.page - 1) * self.size + len(self.items)

    def describe(self) -> str:
        """The "showing 21-40 of 137" line the listing views render."""
        if not self.items:
            return "no results"
        return "showing %d-%d of %d" % (self.start_index, self.end_index, self.total)


def normalize_page_args(page, size) -> tuple:
    try:
        page = max(1, int(page))
    except (TypeError, ValueError):
        page = 1
    try:
        size = int(size)
    except (TypeError, ValueError):
        size = DEFAULT_PAGE_SIZE
    size = max(1, min(MAX_PAGE_SIZE, size))
    return page, size


def paginate(all_items: List, page: int, size: int) -> Page:
    page, size = normalize_page_args(page, size)
    start = (page - 1) * size
    window = all_items[start:start + size]
    return Page(items=window, page=page, size=size, total=len(all_items))


def page_window(current: int, pages: int, span: int = 2):
    """Page numbers to render around the current one, clamped to the range."""
    if pages <= 0:
        return []
    lo = max(1, current - span)
    hi = min(pages, current + span)
    return list(range(lo, hi + 1))


def offset_for(page, size) -> int:
    """Row offset for a page, after both arguments have been normalised."""
    page, size = normalize_page_args(page, size)
    return (page - 1) * size


def empty_page(size: int = DEFAULT_PAGE_SIZE) -> Page:
    """A well-formed empty result, so callers never special-case None."""
    return Page(items=[], page=1, size=size, total=0)
